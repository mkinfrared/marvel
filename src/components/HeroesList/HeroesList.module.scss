.HeroesList {
  height: calc(100vh - 110px);
  margin: 0 auto;
  max-width: 600px;
  padding: 20px 0;
}

.loading {
  align-items: center;
  display: flex;
  justify-content: center;
}
