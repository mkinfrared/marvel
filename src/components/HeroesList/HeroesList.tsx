import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import { FixedSizeList as List } from "react-window";

import useWindowResize from "hooks/useWindowResize";
import { fetchHeroesRequest } from "store/reducers/heroes/reducer";

import css from "./HeroesList.module.scss";
import { HeroesListProps } from "./HeroesList.type";
import Row from "./Row";

const HeroesList = ({ dispatch, heroes, loading }: HeroesListProps) => {
  const [size, setSize] = useState({ x: 0, y: 0 });
  const ref = useRef<HTMLDivElement | null>(null);
  const page = useWindowResize();

  const onEnterHandler = useCallback(() => {
    dispatch(fetchHeroesRequest());
  }, [dispatch]);

  const itemData = useMemo(
    () => ({
      heroes,
      onEnterHandler,
      loading
    }),
    [heroes, onEnterHandler, loading]
  );

  useEffect(() => {
    dispatch(fetchHeroesRequest());
  }, []);

  useEffect(() => {
    if (ref.current) {
      const { clientHeight, clientWidth } = ref.current;

      setSize({ x: clientWidth, y: clientHeight });
    }
  }, [page]);

  const itemSize = page.x < 641 ? 650 : 250;

  return (
    <div className={css.HeroesList} ref={ref}>
      <List
        height={size.y}
        width={size.x}
        itemCount={heroes.length}
        itemSize={itemSize}
        itemData={itemData}
      >
        {Row}
      </List>
    </div>
  );
};

export default React.memo(HeroesList);
