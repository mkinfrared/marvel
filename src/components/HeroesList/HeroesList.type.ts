import { Dispatch } from "@reduxjs/toolkit";

import { fetchHeroesRequest } from "store/reducers/heroes/reducer";
import { Hero } from "store/reducers/heroes/types";

export interface HeroesListProps {
  dispatch: Dispatch<ReturnType<typeof fetchHeroesRequest>>;
  heroes: Hero[];
  loading: boolean;
}
