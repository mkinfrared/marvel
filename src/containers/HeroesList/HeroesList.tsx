import React from "react";
import { useDispatch, useSelector } from "react-redux";

import HeroesListFC from "components/HeroesList";
import {
  selectHeroesList,
  selectHeroesLoading
} from "store/reducers/heroes/selectors";

const HeroesList = () => {
  const dispatch = useDispatch();
  const heroes = useSelector(selectHeroesList);
  const loading = useSelector(selectHeroesLoading);

  return <HeroesListFC dispatch={dispatch} heroes={heroes} loading={loading} />;
};

export default React.memo(HeroesList);
