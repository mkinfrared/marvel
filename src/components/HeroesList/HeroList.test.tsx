import { render } from "@testing-library/react";
import React from "react";

import HeroesList from "./HeroesList";

describe("<HeroesList />", () => {
  const heroes: any[] = [];
  const loading = false;
  const dispatch = jest.fn();
  const props = { loading, heroes, dispatch };
  const Component = <HeroesList {...props} />;

  it("should be defined", () => {
    const { container } = render(Component);

    expect(container).toBeDefined();
  });

  it("match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });
});
