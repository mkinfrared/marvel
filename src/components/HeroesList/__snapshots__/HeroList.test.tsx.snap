// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroesList /> match the snapshot 1`] = `
<div>
  <div
    class="HeroesList"
  >
    <div
      style="position: relative; height: 0px; width: 0px; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 0px; width: 100%;"
      />
    </div>
  </div>
</div>
`;
